ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,605 @@
+* By default `TracePoint` will be enabled for workflow threads that will make sure users are not calling illegal calls,
+  but there is a way to disable for blocks of code.
+* `Ractor`s will be used for state isolation by default assuming decent performance but easy to disable and mostly
+  hidden from users.
+
+These points are explained below.
+
+## Workflow Definition/Usage
+
+Examples below are taken from
+[this Python sample](https://github.com/temporalio/samples-python/tree/main/message_passing/introduction), so not too
+many comments about general non-Ruby behavior are inline.
+
+Example definition:
+
+```ruby
+class CallGreetingServiceActivity < Temporalio::Activity::Definition
+  GREETINGS = {
+    arabic: 'مرحبا بالعالم',
+    chinese: '你好，世界',
+    english: 'Hello, world',
+    french: 'Bonjour, monde',
+    hindi: 'नमस्ते दुनिया',
+    portuguese: 'Olá mundo',
+    spanish: 'Hola mundo'
+  }
+
+  def execute(language)
+    # Simulate a network call
+    sleep(0.2)
+    GREETINGS[language.to_sym]
+  end
+end
+
+class GreetingWorkflow < Temporalio::Workflow::Definition
+  workflow_query_attr :language
+
+  def initialize(name)
+    @greetings = {
+      chinese: '你好，世界',
+      english: 'Hello, world'
+    }
+    @language = :english
+    @mutex = Temporalio::Workflow::Mutex.new
+  end
+
+  def execute
+    Temporalio::Workflow.wait_condition do
+      @approved_for_release && Temporalio::Workflow.all_handlers_finished?
+    end
+    @greetings[@language]
+  end
+
+  workflow_query
+  def languages(input)
+    return CallGreetingServiceActivity::GREETINGS.keys.sort if input[:include_unsupported]
+    @greetings.keys.sort
+  end
+
+  workflow_signal
+  def approve(input)
+    @approved_for_release = true
+    @approver_name = input[:name]
+  end
+
+  workflow_update
+  def set_language(language)
+    previous, @language = @language, language.to_sym
+    previous
+  end
+
+  workflow_update_validator(:set_language)
+  def validate_set_language(language)
+    raise "#{language} not supported" unless @greetings.include?(language.to_sym)
+  end
+
+  workflow_update
+  def set_language_using_activity(language)
+    language = language.to_sym
+    unless @greetings.include?(language)
+      @mutex.synchronize do
+        greeting = Temporalio::Workflow.execute_activity(
+          CallGreetingServiceActivity,
+          language,
+          start_to_close_timeout: 10
+        )
+        unless greeting
+          raise Temporalio::Error::ApplicationError, "Greeting service does not support #{language}"
+        end
+        @greetings[language] = greeting
+      end
+    end
+    previous, @language = @language, language
+    previous
+  end
+end
+```
+
+Example worker:
+
+```ruby
+client = Temporalio::Client.connect('localhost:7233', 'my-namespace')
+worker = Temporalio::Worker.new(
+  client:,
+  task_queue: 'my-task-queue',
+  activities: [GreetingActivity],
+  workflows: [GreetingWorkflow]
+)
+worker.run(shutdown_signals: ['SIGINT'])
```

Comment:
Is `shutdown_signals` a choice that the user needs to make, or is there a good default?

Reply:
This is a choice that a user needs to make. This is an application not a library and it's not neighborly for us to trap signals by default (or have any global side effects by default). This is already in `main` btw and was part of the activity PR.

Reply:
Going to comment here since that PR is in main.
I think this is important since every user will need to set up a `Worker`, and this will be part of the "getting started" scenario, so I have a few questions.
Looks like we default to `[]`.  If this is a choice that every developer needs to make, why is there a default?  What % of users do you think that default will be acceptable for?
Looking at your docs: https://github.com/temporalio/sdk-ruby/blob/67741fcbd9e354b9a8d03e7e923efe8844ccf1ca/temporalio/lib/temporalio/worker.rb#L82

I don't see any guidance.  It's typed as `String | Int` in rbs, and I don't see any validations.  How do users know what values are acceptable or get guidance on what's normal to pass in there?

Reply:
>  If this is a choice that every developer needs to make, why is there a default?

It is not a choice every developer will make. This is a library and many developers will prefer to use a block body or the `Cancellation` token as a way to stop the worker(s). Signals is but one way to stop a worker, by no means the deafult and by no means one that everyone will want. These workers are just things that run how the user wants, we just provide helpers.

> How do users know what values are acceptable or get guidance on what's normal to pass in there in what situations and get feedback if they make a mistake?

How do they know that with `Signal.trap`? That's typed in https://github.com/ruby/rbs/blob/d9000d23c2c1d032a194848562806896c54fd0af/core/signal.rbs#L92-L93 and https://github.com/sorbet/sorbet/blob/0ab8f3c29260f8ecee31aa960de3692419e2d3c3/rbi/core/signal.rbi#L100-L113 as general item (I need to add `Symbol` there it seems). Why are we expecting Temporal to type signals better than RBS and Sorbet?